### PR TITLE
Fix string engine incorrectly skipping identical elements at the same depth level

### PR DESCRIPTION
**Pull request by thibaudcolas**
Interestingly, there was no test for _that_! I've added some. This bug is easily reproducible on http://draftjs-exporter.herokuapp.com/:

string-engine-same-elements

The problem is that the `append_child` method checks if elements are already present by value rather than by reference, so if two elements have equal values (same type, same children, same   attributes) the subsequent ones will be skipped.

I initially changed the check so it would be by reference (`any(c is child for c in elt['children'])`), but that's ± 5% slower in the benchmark so instead I converted the dicts into class instances so the `in` operator looks at references.

### CODE REVIEW

File: draftjs_exporter/engines/string.py [hunk 1]
```diff
@@ -32,54 +32,53 @@ def escape(s):
...
+            elt.children.append(child)
 
     @staticmethod
     def render_attrs(attr):
         return ''.join(sorted([' %s="%s"' % (k, escape(v)) for k, v in attr.items()]))
 
     @staticmethod
     def render_children(children):
-        return ''.join([DOMString.render(c) if isinstance(c, dict) else escape(c) for c in children])
+        return ''.join([DOMString.render(c) if isinstance(c, Elt) else escape(c) for c in children])
```
Comment (loicteixeira) [inline draftjs_exporter/engines/string.py]: Unrelated to the change, but `join` accepts a generator. Therefore instead of passing a list (from the list comprehension), you should be able to pass the generator directly (the comprehension without the square brackets) so the `join` method will iterate the items without the list being fully built in memory.

TL;DR;
Do `return ''.join(DOMString.render(c) if isinstance(c, Elt) else escape(c) for c in children)`

File: draftjs_exporter/engines/string.py [hunk 2]
```diff
@@ -32,54 +32,53 @@ def escape(s):
 }
 
 
+class Elt(object):
+    def __init__(self, type_, attr, children, markup):
```
Comment (loicteixeira) [inline draftjs_exporter/engines/string.py]: When creating `Elt` object, it seems you often have to pass `None` for `attr`, `children` and `markup` so I reckon those should be optional arguments.

In addition, `children` use to always be a `list` with the old `dict` implementation but now it might be `None`. Since in some places the code does `elt.children.append(child)`, if `children` is `None`, it will crash.

Proposed implementation:
```python
def __init__(self, type_, attr=None, children=None, markup=None):
    self.type = type_
    self.attr = attr
    self.children = children if children is not None else []
    self.markup = markup
```

Does it make sense?

Comment (thibaudcolas) [reply draftjs_exporter/engines/string.py]: 👌 `children` was None before when the element is HTML. This is rather hack-ish – I've added a factory method so it's more contained at least.

File: draftjs_exporter/engines/string.py (hunk 1 above)
Comment (thibaudcolas) [reply draftjs_exporter/engines/string.py]: I don't see a performance (speed / memory) gain doing this, can't tell whether this is because the optimisation doesn't make big enough of a difference or because my test setup just isn't stable enough.
